shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs

```diff
@@ -36,97 +36,88 @@ public static string GetTypeDisplayName(object item, bool fullName = true)
+        private static void ProcessTypeName(StringBuilder builder, Type type, bool fullName)
         {
-            var totalArgs = args.Length;
-            if (totalArgs >= startIndex + numberOfArgsToAppend)
+            if (type.GetTypeInfo().IsGenericType)
             {
-                sb.Append("<");
-                for (int i = startIndex; i < startIndex + numberOfArgsToAppend; i++)
-                {
-                    ProcessTypeName(args[i], sb, fullName);
-                    if (i + 1 < startIndex + numberOfArgsToAppend)
-                    {
-                        sb.Append(", ");
-                    }
-                }
-                sb.Append(">");
-            }
-        }
-
-        private static void ProcessTypeName(Type t, StringBuilder sb, bool fullName)
-        {
-            if (t.GetTypeInfo().IsGenericType)
-            {
-                ProcessNestedGenericTypes(t, sb, fullName);
+                ProcessGenericType(builder, type, fullName);
                 return;
             }
-            if (_builtInTypeNames.ContainsKey(t))
+            if (_builtInTypeNames.ContainsKey(type))
             {
-                sb.Append(_builtInTypeNames[t]);
+                builder.Append(_builtInTypeNames[type]);
             }
             else
             {
-                sb.Append(fullName ? t.FullName : t.Name);
+                builder.Append(fullName ? type.FullName : type.Name);
             }
         }
 
-        private static void ProcessNestedGenericTypes(Type t, StringBuilder sb, bool fullName)
+        private static void ProcessGenericType(StringBuilder builder, Type type, bool fullName)
         {
-            var genericFullName = t.GetGenericTypeDefinition().FullName;
-            var genericSimpleName = t.GetGenericTypeDefinition().Name;
-            var parts = genericFullName.Split('+');
-            var genericArguments = t.GetTypeInfo().GenericTypeArguments;
-            var index = 0;
-            var totalParts = parts.Length;
-            if (totalParts == 1)
-            {
-                var part = parts[0];
-                var num = part.IndexOf('`');
-                if (num == -1) return;
+            var genericArguments = type.GetGenericArguments();
+            ProcessNestedGenericType(builder, type, genericArguments, genericArguments.Length, fullName);
+        }
 
-                var name = part.Substring(0, num);
-                var numberOfGenericTypeArgs = int.Parse(part.Substring(num + 1));
-                sb.Append(fullName ? name : genericSimpleName.Substring(0, genericSimpleName.IndexOf('`')));
-                AppendGenericArguments(genericArguments, index, numberOfGenericTypeArgs, sb, fullName);
-                return;
+        private static void ProcessNestedGenericType(StringBuilder builder, Type type, Type[] genericArgs, int limit, bool fullName)
```

Comment:
If we are using helper method, then keep return type int. The returning value should tell how many of args have been used so we don't need to calculate offset by looking at declaring type.